dlt/extract/source.py

```diff
@@ -614,6 +614,15 @@ def root_key(self) -> bool:
         config = RelationalNormalizer.get_normalizer_config(self._schema).get("propagation")
         return config is not None and "root" in config and "_dlt_id" in config["root"] and config["root"]["_dlt_id"] == "_dlt_root_id"
 
+    @property
+    def exhausted(self) -> bool:
+        # check all selected pipes wether one of them is exhausted
+        for resource in self._resources.extracted.values():
+            for item in resource._pipe:
+                if isinstance(item, types.GeneratorType):
```

Comment:
interesting! I'm using `inspect.isgenerator` and it's probably better to stick to that (or you double check if the above catches all cases of generators)